src/module/rules/rule-element/roll-option/rule-element.ts

```diff
@@ -335,13 +335,13 @@ class RollOptionRuleElement extends RuleElementPF2e<RollOptionSchema> {
 
         if (this.mergeable && selection) {
             // Update the items containing rule elements in the merge family
-            const rules = R.groupBy(R.uniq(this.#resolveSuboptions().map((s) => s.rule)), (r) => r.item.id);
-            for (const itemId of Object.keys(rules)) {
+            const rulesByItem = R.groupBy(R.uniq(this.#resolveSuboptions().map((s) => s.rule)), (r) => r.item.id);
+            for (const [itemId, rules] of Object.entries(rulesByItem)) {
                 const item = actor.items.get(itemId, { strict: true });
                 const ruleSources = item.toObject().system.rules;
-                const rollOptionSources = ruleSources.filter(
-                    (_r, index): _r is RollOptionSource => rules[itemId][index]?.sourceIndex === index,
-                );
+                const rollOptionSources = rules
+                    .map((r) => (typeof r.sourceIndex === "number" ? ruleSources[r.sourceIndex] : null))
+                    .filter((r): r is RollOptionSource => r?.key === "RollOption");
```

Comment:
Filter then map?

Reply:
the reason is because it could theoretically fail the array lookup , and we need to get rid of falsy. Might as well type assert while we're at.

Reply:
I guess r as the prop in both might have obscured what's going on. rule vs ruleSource.